### ISSUE TITLE
1.083/1.082 - Array does not understand #byteSize

### ISSUE THREAD

**Issue by MFeldtmann** [jgfoster/Jade.bak#18]
I get more and more this error. Here the dump ( connected against a 3.2.7 Gemstone database). This happens here when I execute a script in a workspace like this:

| useOSTimeZone swap path osName zonename aTimeZone |

swap := true.
useOSTimeZone := true.

useOSTimeZone 
  ifTrue: 
    [ osName := System gemVersionReport at: #osName.
    osName = 'Linux' ifTrue: [ aTimeZone := TimeZone fromLinux ].
    osName = 'SunOS' ifTrue: [ aTimeZone := TimeZone fromSolaris ].
    ].

aTimeZone isNil
  ifTrue:
    [ path := '$GEMSTONE/pub/timezone/etc/zoneinfo/'.
    zonename := 'America/Los_Angeles'.
    aTimeZone := TimeZone fromGemPath: path, zonename.
    ].

swap 
   ifTrue: [TimeZone default become: aTimeZone]
   ifFalse: [TimeZone default: aTimeZone].

TimeZone default installAsCurrentTimeZone.

*************************\* Dolphin Virtual Machine Dump Report ***************************

19:39:55, 17.09.2015: Array does not understand #byteSize

_----> VM Context <----_
Process: {0BCB0004:size 267 words, suspended frame 0BCB0549, priority 5, callbacks 0
last failure 0:nil, FPE mask 3, thread 00000000}
Active Method: SessionManager>>logError:
IP: 006AACA9 (9)
SP: 0BCB0510
BP: 0BCB04E4 (297)
ActiveFrame: {0BCB04E8: cf 0BCB04C9, sp 0BCB0500, bp 0BCB04E4, ip 5, JadeSessionManager(SessionManager)>>logError:}
    receiver: a JadeSessionManager
    arg[0]: a MessageNotUnderstood

New Method: VMLibrary>>dump:path:stackDepth:walkbackDepth:
Message Selector: dump:path:stackDepth:walkbackDepth:

_----> Stack Back Trace <----_
{0BCB04E8: cf 0BCB04C9, sp 0BCB0500, bp 0BCB04E4, ip 5, JadeSessionManager(SessionManager)>>logError:}
    receiver: a JadeSessionManager
    arg[0]: a MessageNotUnderstood

{0BCB04C8: cf 0BCB04A9, sp 0BCB04DC, bp 0BCB04C4, ip 3, JadeSessionManager(SessionManager)>>unhandledException:}
    receiver: a JadeSessionManager
    arg[0]: a MessageNotUnderstood

{0BCB04A8: cf 0BCB0489, sp 0BCB04BC, bp 0BCB04A4, ip 3, JadeSessionManager(SessionManager)>>onUnhandledError:}
    receiver: a JadeSessionManager
    arg[0]: a MessageNotUnderstood

{0BCB0488: cf 0BCB046D, sp 0BCB049C, bp 0BCB0488, ip 5, MessageNotUnderstood(Error)>>defaultAction}
    receiver: a MessageNotUnderstood

{0BCB046C: cf 0BCB0439, sp 0BCB0480, bp 0BCB0454, ip 55, MessageNotUnderstood(Exception)>>_propagateFrom:}
    receiver: a MessageNotUnderstood
    arg[0]: a ExceptionHandler
    stack temp[0]: nil
    stack temp[1]: a ExceptionHandler
    stack temp[2]: nil
    stack temp[3]: a Process(GciSession>>withExplanation:doA: base 0BCB0000 [ACTIVE] in SessionManager>>logError: sp=00000000 ip=4 list=02B60010)
    stack temp[4]: nil

{0BCB0438: cf 0BCB0419, sp 0BCB044C, bp 0BCB0434, ip 6, MessageNotUnderstood(Exception)>>_propagate}
    receiver: a MessageNotUnderstood
    stack temp[0]: nil

{0BCB0418: cf 0BCB03FD, sp 0BCB042C, bp 0BCB0418, ip 11, MessageNotUnderstood(Exception)>>signal}
    receiver: a MessageNotUnderstood

{0BCB03FC: cf 0BCB03D9, sp 0BCB0410, bp 0BCB03F4, ip 12, MessageNotUnderstood class>>receiver:message:}
    receiver: MessageNotUnderstood
    arg[0]: a Array
    arg[1]: Message selector: byteSize arguments: #()

{0BCB03D8: cf 0BCB03B9, sp 0BCB03EC, bp 0BCB03D4, ip 4, Array(Object)>>doesNotUnderstand:}
    receiver: a Array
    arg[0]: Message selector: byteSize arguments: #()

{0BCB03B8: cf 0BCB0391, sp 0BCB03CC, bp 0BCB03B0, ip 3, GciErrSType64_31>>args:}
    receiver: a GciErrSType64_31
    arg[0]: a Array
    stack temp[0]: nil

{0BCB0390: cf 0BCB0349, sp 0BCB03A8, bp 0BCB0364, ip 63, JadeServer>>readExceptionFrom:errorClass:}
    receiver: a JadeServer
    arg[0]: a ReadStream
    arg[1]: GciErrSType64_31
    stack temp[0]: a OopType64
    stack temp[1]: 2116
    stack temp[2]: a OopType64
    stack temp[3]: a OopType64
    stack temp[4]: SecurityError
    stack temp[5]: 'a SecurityError occurred (error 2116), An attempt was made to modify the object ...'
    stack temp[6]: a Array
    stack temp[7]: '1 [] in  ExecBlock1 (JadeServer64bit3x) >> reportErrorOnStream:fromEvaluationOf:...'
    stack temp[8]: a GciErrSType64_31

{0BCB0348: cf 0BCB0301, sp 0BCB035C, bp 0BCB031C, ip 371, JadeServer>>readObjectFrom:errorClass:}
    receiver: a JadeServer
    arg[0]: a ReadStream
    arg[1]: GciErrSType64_31
    stack temp[0]: 13
    stack temp[1]: nil
    stack temp[2]: nil
    stack temp[3]: nil
    stack temp[4]: nil
    stack temp[5]: nil
    stack temp[6]: nil
    stack temp[7]: nil
    stack temp[8]: nil
    env temp[0]: nil
    env temp[1]: nil

{0BCB0300: cf 0BCB02C5, sp 0BCB0314, bp 0BCB02E0, ip 72, GciSession>>_usingSocketSend:to:withAll:}
    receiver: a GciSession
    arg[0]: executeString:fromContext:
    arg[1]: a OopType64
    arg[2]: a Array
    stack temp[0]: a ReadStream
    stack temp[1]: a Process(BlockClosure>>newProcessWithArguments: base 0C390000 frame=nil)
    stack temp[2]: 1532
    stack temp[3]: a ByteArray
    stack temp[4]: a JadeServer

{0BCB02C4: cf 0BCB02A1, sp 0BCB02D8, bp 0BCB02BC, ip 20, GciSession>>_executeString:fromContextOop:}
    receiver: a GciSession
    arg[0]: '| useOSTimeZone swap path osName zonename aTimeZone |\0A\0A\swap := true.\0A\useOSTimeZo...'
    arg[1]: a OopType64

{0BCB02A0: cf 0BCB027D, sp 0BCB02B4, bp 0BCB0298, ip 15, [] in GciSession>>executeString:fromContext:}
    receiver: a GciSession

{0BCB027C: cf 0BCB025D, sp 0BCB0290, bp 0BCB0278, ip 11, [] in GciSession>>returningResultOrErrorDo:}
    receiver: nil

{0BCB025C: cf 0BCB0241, sp 0BCB0270, bp 0BCB025C, ip 13, ExceptionHandler(ExceptionHandlerAbstract)>>markAndTry}
    receiver: a ExceptionHandler

{0BCB0240: cf 0BCB0219, sp 0BCB0254, bp 0BCB023C, ip 22, [] in ExceptionHandler(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandler

{0BCB0218: cf 0BCB01F9, sp 0BCB022C, bp 0BCB0214, ip 18, BlockClosure>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:

{0BCB01F8: cf 0BCB01D5, sp 0BCB020C, bp 0BCB01F0, ip 3, BlockClosure>>ensure:}
    receiver: [] @ 17 in ExceptionHandlerAbstract>>try:
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:
    stack temp[0]: nil

{0BCB01D4: cf 0BCB01AD, sp 0BCB01E8, bp 0BCB01C8, ip 44, ExceptionHandler(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandler
    arg[0]: [] @ 9 in GciSession>>returningResultOrErrorDo:
    stack temp[0]: a ExceptionHandler
    stack temp[1]: a Process(GciSession>>withExplanation:doA: base 0BCB0000 [ACTIVE] in SessionManager>>logError: sp=00000000 ip=4 list=02B60010)
    env temp[0]: nil

{0BCB01AC: cf 0BCB0189, sp 0BCB01C0, bp 0BCB01A4, ip 7, BlockClosure>>on:do:}
    receiver: [] @ 9 in GciSession>>returningResultOrErrorDo:
    arg[0]: GciError
    arg[1]: [] @ 20 in GciSession>>returningResultOrErrorDo:

{0BCB0188: cf 0BCB0169, sp 0BCB019C, bp 0BCB0184, ip 29, GciSession>>returningResultOrErrorDo:}
    receiver: a GciSession
    arg[0]: [] @ 11 in GciSession>>executeString:fromContext:

{0BCB0168: cf 0BCB013D, sp 0BCB017C, bp 0BCB0158, ip 5, GciSession>>handlingClientForwarderSendDo:}
    receiver: a GciSession
    arg[0]: [] @ 11 in GciSession>>executeString:fromContext:
    stack temp[0]: [] @ 11 in GciSession>>executeString:fromContext:
    stack temp[1]: nil
    stack temp[2]: nil

{0BCB013C: cf 0BCB011D, sp 0BCB0150, bp 0BCB0138, ip 16, [] in GciSession>>withExplanation:doA:}
    receiver: a GciSession

{0BCB011C: cf 0BCB0101, sp 0BCB0130, bp 0BCB011C, ip 13, ExceptionHandler(ExceptionHandlerAbstract)>>markAndTry}
    receiver: a ExceptionHandler

{0BCB0100: cf 0BCB00D9, sp 0BCB0114, bp 0BCB00FC, ip 22, [] in ExceptionHandler(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandler

{0BCB00D8: cf 0BCB00B9, sp 0BCB00EC, bp 0BCB00D4, ip 18, BlockClosure>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:

{0BCB00B8: cf 0BCB0095, sp 0BCB00CC, bp 0BCB00B0, ip 3, BlockClosure>>ensure:}
    receiver: [] @ 17 in ExceptionHandlerAbstract>>try:
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:
    stack temp[0]: nil

{0BCB0094: cf 0BCB006D, sp 0BCB00A8, bp 0BCB0088, ip 44, ExceptionHandler(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandler
    arg[0]: [] @ 14 in GciSession>>withExplanation:doA:
    stack temp[0]: nil
    stack temp[1]: a Process(GciSession>>withExplanation:doA: base 0BCB0000 [ACTIVE] in SessionManager>>logError: sp=00000000 ip=4 list=02B60010)
    env temp[0]: nil

{0BCB006C: cf 0BCB0049, sp 0BCB0080, bp 0BCB0064, ip 7, BlockClosure>>on:do:}
    receiver: [] @ 14 in GciSession>>withExplanation:doA:
    arg[0]: ProcessTermination
    arg[1]: [] @ 13 in BlockClosure>>newProcess

{0BCB0048: cf 00000001, sp 0BCB005C, bp 0BCB0048, ip 17, [] in BlockClosure>>newProcess}
    receiver: [] @ 14 in GciSession>>withExplanation:doA:

<Bottom of stack>

**Comment by MFeldtmann** [comment]
Today I got this error in another situation. I selected a class and wanted to add missing accessing methods. In the runtime I get the same error as mentioned above. 

In the development version I get:

a LookupError occurred (error 2096), reason:assocErrNoElementsDetected, detect: was sent to a collection, but no elements of the collection anIdentitySet( Array, Association, IdentityBag, Behavior, Boolean, Character, Class, Collection, CompiledMethod, DecimalFloat, Integer, InvariantArray, InvariantString, Magnitude, Number, Object, PositionableStream, ReadStream, OldRepository, GsObjectSecurityPolicy, SequenceableCollection, IdentitySet, SmallInteger, Stream, String, System, UndefinedObject, UserProfile, AbstractUserProfileSet, WriteStream, Fraction, SymbolList, AbstractCollisionBucket, KeyValueDictionary, IntegerKeyValueDictionary, StringKeyValueDictionary, CharacterCollection, JapaneseString, EUCString, InvariantEUCString, EUCSymbol, AbstractCharacter, JISCharacter, ClusterBucket, ClusterBucketArray, ClassHistory, RcQueue, StackSegment, ...) satisfied the block anExecBlock1.

The error is thrown in:

JadeServer64bit32>>sbClassFrom: anOrderedCollection

```
| selectedClassName selectedClassOop set myClass |
selectedClassName := (anOrderedCollection removeFirst subStrings: Character space) first.
selectedClassOop := anOrderedCollection removeFirst asNumber.
set := IdentitySet new.
self symbolList do: [:eachDict | 
    eachDict do: [:eachGlobal |
        eachGlobal isBehavior ifTrue: [
            set addAll: (self historyOf: eachGlobal).
        ].
    ].
].
```

-> in the following statement the error is thrown
    myClass := set detect: [:each | (self oopOf: each) = selectedClassOop].

.....

**Comment by MFeldtmann** [comment]
I closed this issue - I opened it again.

The error is thrown in the follown method (detect: -statement) and in anOrderedCollection is only one element: 'instance tab':

JadeServer64bit32>>sbClassFrom: anOrderedCollection

```
| selectedClassName selectedClassOop set myClass |
selectedClassName := (anOrderedCollection removeFirst subStrings: Character space) first.
selectedClassOop := anOrderedCollection removeFirst asNumber.
set := IdentitySet new.
self symbolList do: [:eachDict | 
    eachDict do: [:eachGlobal |
        eachGlobal isBehavior ifTrue: [
            set addAll: (self historyOf: eachGlobal).
        ].
    ].
].
myClass := set detect: [:each | (self oopOf: each) = selectedClassOop].
myClass name asString = selectedClassName ifFalse: [self error: 'Class not found!'].
anOrderedCollection removeFirst = 'classTab' ifTrue: [myClass := myClass class].
^myClass.
```

**Comment by MFeldtmann** [comment]
I have to say, that I added an attribute to this class just before that and I still can not create missing accessors.

**Comment by MFeldtmann** [comment]
I did this in the "Hierarchy Mode". In the "Classes"-Mode I get a totally other error, BUT I see two entries of that class - but none of them can be selected.

**Comment by jgfoster** [comment]
This is odd. Starting with the first stack, I see the error "a SecurityError occurred (error 2116), An attempt was made to modify the object ..." and this is what I got when I tried to set the TimeZone as DataCurator rather than as SystemUser. Of course, the error should be reported properly so it isn't enough just to say "use the right login"!

Next, I see that GciSession>>_usingSocketSend:to:withAll: is called by GciSession>>_executeString:fromContextOop:. The first thing this suggests is that on the 'Advanced' tab on the Jade Login Shell the 'Open new socket to improve performance (experimental)' is checked. If so, then you should uncheck this and try again. 

I'm confused, however, on how this even could happen. In Jade 1.083 the method GciSession>>_usingSocketSend:to:withAll: does not exist. It was added to the package GemStone Session 0.225 on 31-July-2015 while 1.083 uses GemStone Session 0.224 from 17-June-2015. Since 1.083 was released on 28-July-2015 as https://github.com/jgfoster/Jade/commit/92ce9047ae906e9539ae4030ece404eecb297f0d and does not have the method in your stack, it seems like you are running something I haven't published to GitHub. Did I send you a private build (perhaps to fix some other problem)?

**Comment by MFeldtmann** [comment]
Hello James,

I get this Array does not understand byteSize error on a very regular
base now (since around August - i started from holidy around
09/08/2015). Regular means very often and in general I can then exit the
jade IDE because it feels _very_ strange then (busy cursor).

I wanted to get to know (today), where this error happens and then I
downloaded the current github stuff as a zip and started this within
Dolphin - actually I did it today and then I noticed, that Dolphin 1.083
simply gets an error (actually an exception) somewhere but does not
return this error, but the Array .... error. I knew, that I did not had
the access rights to execute this code, but I wanted to know, why I do
not get a correct error message and all this Array ... error.

To summarize: If I use the 1.083 and execute the initializeTimezone code
I do not get the correct error message - but that strange one above.

Now I work (for some hours) from within the IDE and I noticed that I get
this Array bytesize error whenever an exception occured within Dolphin.
I even can not debug or single step from within the Dolphin IDE - I
always get exceptions.

All tests were done either with 3.2.7 or (later) with 3.2.9.

Very, Very strange here ...

Marten

Am 18.09.2015 um 21:22 schrieb James Foster:

**Comment by jgfoster** [comment]
Based on the above analysis, I'm working from the assumption that the error "Array does not understand #byteSize" is a side-effect of running something other than the Jade runtime version 1.083. From the original stack, it appears that Jade is attempting to report another error. In the first case, it was an authorization error (and correct, since the default TimeZone can only be set as SystemUser). The subsequent error(s) appear to be from attempting to add a missing accessor. It appears that Jade is unable to find a class that it thinks should exist, apparently after adding an instance variable.

If the Jade login was done with the 'Socket' checked and the error reporting improves when it is not checked, then I'd like to close this issue. If the lost class issue reproduces then I'd like to have another issue opened with a description of the series of steps taken to cause the confusion.

**Comment by jgfoster** [comment]
Marten,
On the Jade Login Shell select the Advanced tab and uncheck the box next to "Open new socket ...". After this login and try again.
James

**Comment by MFeldtmann** [comment]
Hey, after unchecking the option now I get the original error message.

Thats good !

Marten

Am 18.09.2015 um 22:30 schrieb James Foster:

**Comment by jgfoster** [comment]
This is tied to the experimental socket protocol.
